lib/lattice_observer/observed/decay.ex

```diff
@@ -6,18 +6,29 @@ defmodule LatticeObserver.Observed.Decay do
   # Based on the decay rate in the lattice parameter, hosts should
   # age out in descending order of health status
   def age_hosts(l = %Lattice{}, event_time) do
+    {fully_decayed, hosts} =
+      age_hosts(l.hosts, event_time, l.parameters.host_status_decay_rate_seconds)
+
+    # Remove each fully decayed host and its resources from the lattice
+    lattice =
+      List.foldl(fully_decayed, l, fn {hk, _host}, lattice ->
+        LatticeObserver.Observed.EventProcessor.remove_host(lattice, hk)
+      end)
+
+    # Return the lattice with the list of hosts with updated decay states
     %Lattice{
-      l
-      | hosts: age_hosts(l.hosts, event_time, l.parameters.host_status_decay_rate_seconds)
+      lattice
+      | hosts: hosts |> Enum.into(%{})
     }
   end
 
-  @spec age_hosts(Lattice.hostmap(), DateTime.t(), Integer.t()) :: Lattice.hostmap()
-  def age_hosts(hosts, event_time, decay_rate) do
+  # Returns a tuple with two elements, the first is a list of hosts that have fully
+  # decayed and are removed from the lattice, the second is the new list of hosts
+  @spec age_hosts(Lattice.hostmap(), DateTime.t(), Integer.t()) :: {list(), list()}
+  defp age_hosts(hosts, event_time, decay_rate) do
```

Comment:
This reads as imperative, but it's functional (as all things should be 😉). How about renaming this to something like `partition_decayed_hosts`?